README.md
# react-xarrows

## introduction

Draw arrows between components in React!

[![npm version](https://badge.fury.io/js/react-xarrows.svg)](https://github.com/Eliav2/react-xarrows)
[![downloads](https://img.shields.io/npm/dw/react-xarrows)](https://www.npmjs.com/package/react-xarrows)
[![issues](https://img.shields.io/github/issues/Eliav2/react-xarrows)](https://github.com/Eliav2/react-xarrows/issues)
[![licence](https://img.shields.io/npm/l/react-xarrows)](https://github.com/Eliav2/react-xarrows/blob/master/LICENSE)

#### main features

- Connect between elements by passing a ref or an id for startElement and endElement.
- Automatic anchoring based on smallest path.
- Fast algorithm to find path and to adjust canvas.
- Customization is easy but flexible

found a problem? not a problem! post a new issue([here](https://github.com/Eliav2/react-xarrows/issues)).

liked my work? please star [this repo](https://github.com/Eliav2/react-xarrows).

## installation

with npm `npm install react-xarrows`.
(or `yarn add react-xarrows`)

## Examples

#### Demos

[see here!](https://codesandbox.io/embed/github/Eliav2/react-xarrows/tree/master/examples?fontsize=14&hidenavigation=1&theme=dark)
codebox of few examples(in this repo at [/examples](./examples)).

![react-xarrow-picture-1 4 2](https://user-images.githubusercontent.com/47307889/87698325-facfc480-c79b-11ea-976a-dbad0ecd9b48.png)

see this interactive example: <https://lwwwp.csb.app/CustomizeArrow>

[comment]: <> (![react-xarrows-v1 6]&#40;https://user-images.githubusercontent.com/47307889/113949468-070f1c80-9818-11eb-90e6-ddc6d814b912.gif&#41;)

<img src="https://user-images.githubusercontent.com/47307889/113949468-070f1c80-9818-11eb-90e6-ddc6d814b912.gif" width="650px"/>

### simple example:

```jsx
import React, {useRef} from "react";
import Xarrow from "react-xarrows";

const boxStyle = {
    border: "grey solid 2px",
    borderRadius: "10px",
    padding: "5px",
};

function SimpleExample() {
    const box1Ref = useRef(null);
    return (
        <div style={{display: "flex", justifyContent: "space-evenly", width: "100%"}}>
            <div ref={box1Ref} style={boxStyle}>
                hey
            </div>
            <p id="elem2" style={boxStyle}>
                hey2
            </p>
            <Xarrow
                start={box1Ref} //can be react ref
                end="elem2" //or an id
            />
        </div>
    );
}

export default SimpleExample;
```

## Usage

react-xarrows does not renders automatically if one of the connected elements is rendered. You have to manually trigger
an update on the arrows whenever one of the connected elements renders(possibaly by trigger update on the parent of the
arrows) ,this is because the Xarrow component does not have any control or awareness of the connected elements. in
addition.
**this is planned to be changed in react-xarrows v2.**

### Contributing

Want a feature that is not supported? found a bug?\
no need to clone the repo and set up the dev environment anymore!\
here's a ready to use development environment with a click of a button(patience, it takes about a minute to setup):

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/Eliav2/react-xarrows/blob/master/src/index.tsx)

this will set up environment that will clone react-xarrow master,and will link the code from the src to the examples,
and will start examples,with typescript watch process that will recompile when any change is made.\
this means that any code changes in src/index.tsx will immediately be reflected to the running example at port 3000!
(add console.log("test") line and see!)\
to reproduce this dev env on your local machine git clone and follow same commands as in [gitpod.yml](./.gitpod.yml).

if you made an improvement that is relevant for most users, you can quickly submit a pull request.  
Please always pull request from and into dev branch -
here's [Gitpod](https://gitpod.io/#https://github.com/Eliav2/react-xarrows/blob/dev/src/index.tsx)

### react-xarrows v2

v2 is on its way. want to contribute and participate in plannig the next react architecture for react-xarrows? see
discussion [here](https://github.com/Eliav2/react-xarrows/discussions/53)!

### API

to see full typescript definition see [types.ts](./src/types.ts) file.

here's a summary of the all the available props:

**Properties**|**Description**|**default value**|**type**
:-----:|:-----:|:-----:|:-----:
[start](#refs)|ref to start element|none(Required!)|string/ReactRef
[end](#refs)|ref to end element|none(Required!)|string/ReactRef
[startAnchor](#anchors)|from which side the arrow should start from start element| 'auto'|string/object/array
[endAnchor](#anchors)|at which side the arrow should end at end element| 'auto'|string/object/array
[label](#label)|optional labels| null|string/array
[color](#colors)|color of Xarrow(all parts)| 'CornflowerBlue'|string
[lineColor](#colors)|color of the line| null|string
[headColor](#colors)|color of the head| null|string
[tailColor](#colors)|color of the tail| null|string
[strokeWidth](#widths)|thickness of Xarrow(all parts)|4|number
[headSize](#widths)|thickness of head(relative to strokeWidth)|6|number
[tailSize](#widths)|thickness of tail(relative to strokeWidth)|6|number
[path](#path)|path drawing style| 'smooth'|string
[curveness](#curveness)|how much the line curveness when path='smooth'| 0.8|number
[gridBreak](#gridBreak)|where the line breaks in path='grid'| 0.5|number
[dashness](#dashness)|should the line be dashed| false|boolean/object
[showHead](#shows)|show the arrow head?| true|boolean
[showTail](#shows)|show the arrow tail?| false|boolean
[showXarrow](#shows)|show Xarrow?| true|boolean
[animateDrawing](#animateDrawing)|animate drawing when arrow mounts?| false|boolean/object
[headShape](#customsvgs)|shape of the arrow head| 'arrow1'|string/object
[tailShape](#customsvgs)|shape of the arrow tail|'arrow1'|string/object

<details>

<summary markdown='span'>Advanced Props</summary>

[see details](#advancedCustom)

**Properties**|**Description**|**default value**|**type**
:-----:|:-----:|:-----:|:-----:
passProps|properties which will be pased to arrowBody,arrowHead,arrowTail| {}|object
SVGcanvasProps|properties which will be passed to svgCanvas| {}|object
arrowBodyProps|properties which will be passed to arrowBody| {}|object
arrowHeadProps|properties which will be passed to arrowHead| {}|object
arrowTailProps|properties which will be passed to arrowTail| {}|object
divContainerProps|properties which will be passed to divContainer| {}|object
SVGcanvasStyle|style properties which will be passed svgCanvas|0|object
divContainerStyle|style properties which will be passed divContainer| false|object
_extendSVGcanvas|extend svgCanas at all sides|0|number
_debug|show debug elements|0|boolean
_cpx1Offset|offset control point 1 x|0|number
_cpy1Offset|offset control point 1 y|0|number
_cpx2Offset|offset control point 2 x|0|number
_cpy2Offset|offset control point 2 x|0|number

</details>

##### API flexibility

This API is built in such way that most props can accept different types. you can keep things simple or provide more
custom props for more custom behavior - the API except both(see [`startAnchor`](#anchors) or `label`
properties for good examples)
.<br/>
see typescript types above for detailed descriptions of what type excepts every prop.

## Properties

This documentation is examples driven.\
The examples is sorted from the most common use case to the most custom one.

<a name="refs"></a>

<details>

<summary markdown='span'> 'start' and 'end' </summary>

_required_\
can be a reference to a react ref to html element or string - an id of a DOM element.

examples:

- `start="myid"` - `myid` is id of a dom element.
- `start={myRef}` -  `myRef` is a react ref.

</details>

<a name="anchors"></a>

<details>

<summary markdown='span'> 'startAnchor' and 'endAnchor' </summary>

specify what anchors are allowed. can be a string/object/array.

type:

```typescript
export type anchorType = anchorPositionType | anchorCustomPositionType;
```

#### simple usage:

type:

```typescript
export const tAnchorEdge = ['middle', 'left', 'right', 'top', 'bottom', 'auto'] as const;
export type anchorPositionType = typeof tAnchorEdge[number];
```

one of `"auto" | "middle" | "left" | "right" | "top" | "bottom"`
`auto` will choose automatically the path with the smallest length.

example:

- `endAnchor="middle"` will set the anchor of the end of the line to the middle of the end element.

#### custom usage:

type:

```typescript
export type anchorCustomPositionType = {
    position: anchorPositionType;
    offset: { rightness?: number; bottomness?: number };
};
```

you can offset the anchor from normal positioning. NOTE: breaking changes in naming in v2.

example:

- `endAnchor= { position: "auto", offset: { rightness: 20 } }` will choose automatic anchoring for end anchor but will
  offset it 20 pixels to the right after normal positioning.

if list is provided - the minimal length anchors will be chosen from the list. example:

- `endAnchor= ["right", {position: "left", offset: {bottomness: -10}}]` only right and left anchors will be allowed for
  endAnchor, and if the left side connected then it will be offset 10 pixels up.

</details>

<a name="label"></a>

<details>

<summary markdown='span'> label </summary>

you can place up to 3 labels. see examples

- ```label="middleLabel"``` - middle label
- ```label=<div style={{ fontSize: "1.3em", fontFamily: "fantasy", fontStyle: "italic" }}>styled middle label</div>``` -
  custom middle label
- ```label={{ start:"I'm start label",middle: "middleLabel",end:<div style={{ fontSize: "1.3em", fontFamily: "fantasy", fontStyle: "italic" }}>big end label</div> }}```
  start and middle label and custom end label
- ```label={{ 30: "label" }}``` - label 30 pixels relative to the start of the line
- ```label={{ 30%: "label" }}``` - label on the 30% of the total length from the start of the line
- ```label={{ 30%50: "label" }}``` - 30% of the total length, and 50 more pixels over the line

'start','middle',and 'end' labels have some predefined styles. custom labels (like `30%50: "label"`) will have the same
style as middle label. if you dont like the style of 'start' and 'end' use 2% for start and 98% for end. make sure to
stick between 0-100% in custom labels

you can combine all the different possible labels and can place as many as you want.

</details>

<a name="colors"></a>

<details>

<summary markdown='span'> color,lineColor and headColor and tailColor </summary>


`color` defines color to the entire arrow. lineColor,headColor and tailColor will override color specifically for
line,tail or head. examples:

- `color="red"` will change the color of the arrow to red(body and head).
- `headColor="red"` will change only the color of the head to red.
- `tailColor="red"` will change only the color of the tail to red.
- `lineColor="red"` will change only the color of the body to red.

</details>

<a name="widths"></a>

<details>

<summary markdown='span'>strokeWidth and headSize and tailSize</summary>

strokeWidth defines the thickness of the entire arrow. headSize and tailSize defines how big will be the head or tail
relative to the strokeWidth. examples:

- `strokeWidth={15}` will make the arrow more thick(body and head).
- `headSize={15}` will make the head of the arrow more thick(relative to strokeWidth as well).
- `tailSize={15}` will make arrow's tail thicker.

</details>

<a name="path"></a>

<details>

<summary markdown='span'>path</summary>

`path` can be one of: `"smooth" | "grid" | "straight"`, and it controls the path arrow is drawn, exactly how their name
suggest. examples:

- `path={"grid"}` will draw the line in sharp curves(90 degrees) like grid.

</details>

<a name="curveness"></a>

<details>

<summary markdown='span'>curveness</summary>

defines how much the lines curve. makes a difference only in `path='smooth'`. examples:

- `curveness={false}` will make the line straight without curves(exactly like path='straight').
- `curveness={true}` will choose default values of curveness.
- `curveness={2}` will make Xarrow extra curved.

</details>

<a name="gridBreak"></a>

<details>

<summary markdown='span'>gridBreak</summary>

defines where the line will break when `path='grid'`. value should be a number from 0 to 1.

examples:

- `gridBreak={0.2}` the line would break closer to start element(20% of the path instead of 50%).

</details>

<a name="dashness"></a>

<details>

<summary markdown='span'>dashness</summary>


can make the arrow dashed and can even animate. if true default values(for dashness) are chosen. if object is passed
then default values are chosen except what passed. examples:

- `dashness={true}` will make the line of the arrow to be dashed.
- `dashness={{ strokeLen: 10, nonStrokeLen: 15, animation: -2 }}` will make a custom looking dashness.

</details>

<a name="shows"></a>

<details>

<summary markdown='span'>showHead, showTail and showXarrow</summary>

`showXarrow`: show or not show Xarrow? (can be used to restart the drawing animation)
`showHead`: show or not the arrow head?
`showTail`: show or not the arrow tail?

- `showXarrow={false}` - will hide (unmount) Xarrow and his labels.
- `showHead={false}` - will hide the arrow head.

</details>

<a name="animateDrawing"></a>

<details>

<summary markdown='span'>animateDrawing</summary>


can animate the drawing of the arrow using svg animation. type: boolean|number. if true animation duration is 1s. if
number is passed then animation duration is number's value in seconds. examples:

- `animateDrawing` will animate the drawing of the arrow in 1 second.
- `animateDrawing={5}` will animate the drawing of the arrow in 5 seconds.
- `animateDrawing={0.1}` will animate the drawing of the arrow in 100 milliseconds.

</details>

<a name="customsvgs"></a>

<details>

<summary markdown='span'> custom svg arrows - headShape and tailShape</summary>


[//]: # (todo: add svg custom shapes docs!)

new feature! you can customize the svg edges (head or tail) of the arrow. you can use predefined svg by passing
string,one of `"arrow1" | "circle" | "heart"`

#### simple usage:

_headShape type:string_


<table>
  <tr>
    <th>Code</th>
    <th>Result</th>
  </tr> 
  <tr>
  <td>

```jsx
<xarrow headShape='circle'/>
```

  </td>
  <td> 

![img_1.png](images/fillCircle.png)
  </td>
  </tr>
  <tr>
  <td>

```jsx
<xarrow headShape='circle'
        arrowHeadProps={"fill": "transparent",
            "strokeWidth": "0.1",
            "stroke": "CornflowerBlue"}
/>
```

  </td>
  <td> 

![img_1.png](images/emptyCircle.png)
  </td>
  </tr>
  <tr>
  <td>

```jsx
<xarrow headShape='heart'/>
```

  </td>
  <td> 

![img.png](images/heart.png)
  </td>
  </tr>
</table>

you can import `arrowShapes` which is object contains all predefined svg shapes.

<details>

<summary markdown='span'>custom usage</summary> 

you can also pass _your own_ svg shapes:

```typescript
headShapeType = {
    svgElem: T
:
'circle' | 'ellipse' | 'line' | 'path' | 'polygon' | 'polyline' | 'rect';
svgProps ? : JSX.IntrinsicElements[T];
offsetForward ? : number;
}
;
```

for example, you can pass the following object, and it will be exactly equivalent to passing `'arrow1'`:

```js
headShape = {
    svgElem: 'path',
    svgProps: {d: `M 0 0 L 1 0.5 L 0 1 L 0.25 0.5 z`},
    offsetForward: 0.25
}
```

`svgElem` - an svg element like `path` or `circle`.  
`svgProps` - props that will be passed to the svg element.
`offsetForward` - how much to offset tht line into the svg element(from 0 to 1). normally the line will connect to the
start of the svgElem. for example in case of the default arrow you want the line to enter 25% into the svgElem.

don't forget about `arrowHeadProps` and `arrowTailProps` in case you want to use default shape but custom svg props.

**in case you pass a custom svg element:** currently you have to adjust the path to start from 0,0 and to be at size box
1x1 in order to make the custom shape look like the default shapes in size, in later versions it is planned to support
automatic adjustment using getBBox() function.

</details>

</details>

### advanced customization

<a name="advancedCustom"></a>

<details>

### passing props

The xarrow is fully customizable, and you can pass props to any part of the component. if unlisted(unknown) property is
passed to xarrow so by default it'll be passed down to `divContainer`.

#### passProps

you can pass properties to visible parts(body and head) of the arrow (such event handlers and much more). this supposed
to be enough for most cases. examples:

- `passProps= {{onClick: () => console.log("xarrow clicked!")}}` - now the arrow will console log a message when
  clicked.
- `passProps= {{cursor: "pointer"}}` - now the cursor will change to pointer style when hovering over Xarrow.
- `passProps= {{pointerEvents: "none"}}` - now the user cannot interact with Xarrow via mouse events.

### advanced customization

The properties below can be used to customize the arrow even farther:

#### arrowBodyProps, arrowHeadProps, SVGcanvasProps, divContainerProps

![image](https://user-images.githubusercontent.com/47307889/95031511-09ed5100-06bf-11eb-95a3-4cdc8d0be0ad.png)

if you wish you can pass props specific part of the component.

- `divContainerProps` - the container which contains the SVG canvas, and the optional labels elements. It takes no
  place, and located where you normaly placed him in the elements tree(no offset). The SVGcanvas and the labels will be
  placed in a offset from this div.
- `SVGcanvasProps` - the svg canvas which contains arrow head and body.
- `arrowBodyProps` - the body of the arrow
- `arrowHeadProps` - the arrow head.

Note that `arrowBody` and `arrowHead` receives props of svg path element, `SVGcanvas` receives props of svg element,
and `divContainerProps` of a div element.

examples:

- `arrowHead = {onClick: () => console.log("head clicked!")}` - now only the head will console log a message when
  clicked.

##### SVGcanvasStyle, divContainerStyle

if you wish to pass style to divContainer or SVGcanvas use `SVGcanvasStyle`,`divContainerStyle` and not `SVGcanvasProps`
,`divContainerProps` to not override existing style.

[//]: # (todo: check if width and height can't be automatically determined)

##### _extendSVGcanvas

will extend the svg canvas at all sides. can be useful if for some reason the arrow is cut though to small svg canvas(
should be used in advanced custom arrows). example: `_extendSVGcanvas = {30}` - will extend svg canvas in all sides by
30 pixels.

##### _cpx1Offset,_cpy1Offset,_cpx2Offset,_cpy2Offset

now you can manipulate and offset the control points of the arrow. this way you can control how the line curves. check
out the interactive codesandbox, set _debug to true and play with these properties.

</details>

## Versions

See [CHANGELOG.md](./CHANGELOG.md) in this repo.



<style>
details {
    border: 1px solid #aaa;
    border-radius: 4px;
    padding: .5em .5em 0;
    margin: 1em 0;
}
</style>


CHANGELOG.md
## next
- many many changes
changes that i will forget:
- `gridBreak` type changed from `number` to `string`

## v1.7.0:

big update!

**performance improvements**:

- much faster and optimized render cycle.

**feature update**:

- creation animation added (see `animateDrawing`,`showXarrow`)
- custom svg now supported,added `circle` and `heart` shapes just for examples.

**new properties**:

- `gridBreak`
- `showXarrow`
- `animateDrawing`
- `headShape`
- `tailShape`

Docs improved!

## v1.6.0:

- proptypes - typescript types issues fixed.
- now all pointer events enabled in the arrowhead by default ([#50](https://github.com/Eliav2/react-xarrows/issues/50))
- added properties: `SVGcanvasStyle` and `divContainerStyle` ([#42])(https://github.com/Eliav2/react-xarrows/issues/42)
- examples: bugfixes(playground) and examples update.
- a new feature! tail support!  new properties: `showHead`,`showTail`,`tailSize`,`tailColor`,`arrowTailProps` .
- renamed `extendSVGcanvas` to `_extendSVGcanvas`
- advanced customization support: newProperties: `_cpx1Offset`,`_cpy1Offset`,`_cpx2Offset`,`_cpy2Offset`.
- bugfix: in grid mode in some cases exceeded canvas.
- 1.6.1 - typescript fix: optional properties were require

## 1.5.0

- properties that are completely removed and no longer supported: `consoleWarning`, `monitorDOMchanges`
  and `registerEvents`.
- issues related to jest and other issues has fixed.
- dependencies changed: prop-types added in lodash replaced with lodash.isequal and lodash.pick.
- v1.5.1 - minor code and docs updates. [DEPRECATED]
- v1.5.2 - debug element was published by mistake at v1.5.1.

improved documentation:

- added CHANGELOG.md file.
- added TODO.md file.
- docs in the README.md file updated

## 1.4.0: features update

- `label` property API changed.
- `path` property added

## 1.3.0: bug fixes and features update

- now `startAnchor` and `endAnchor` can be offset from normal position. see `anchorCustomPositionType` type in types
  declaration to see how to offset anchors.
- a new powerful feature: `passProps` - now its possible to pass methods (such event handlers) or attributes to the
  inner components of Xarrow.
- `consoleWarning` default prop changed to `false`.
- bug fixes and inner optimizations(arrow head implemnted now purly using path element without marker elements).## 1.2.0
  added support for javascript projects that imported the lib locally. many changes to the repo folders structure.

## 1.2.0 bug fixes and code structure changed

The lib not worked properly on js projects(only worked well on ts). now working well on both.

- v1.2.1 - minor bug fix (intellij suggestinons did not apear)
- v1.2.2 - bug fixes(1#changing anchors refs without remounting broke the arrow. 2#other minors)

## 1.1.0 bug fixes and features update

props added:`label`, `dashness` and `advanced`.
`arrowStyle` removed and all his contained properties flattened to be props of xarrow directly. `strokeColor` renamed
to `lineColor`.

- v1.1.1 - bug fix now labels not exceed the svg canvas. the headArrow is calcualted now . this means the line ends at
  the start at the arrow - and this is more natural looking(especially at large headarrows).
- v1.1.2 - bug fix. (the first arrow fixed the headarrow style for all next comming arrows)
- v1.1.3 - an entirely new algorithm to calculate arrow path and curveness. now the arrow acting "smarter". this include
  bug fixes,improvements and some adjustments. monitorDOMchanges prop default changed to true.
- v1.1.4 - bug fixes, calculation optimizations, and smart svg canvas size adjusment.
- v1.1.5 - optimazed calculations and label positioning. (Buzier curve extrema point are calculated now using
  derivatives and not by interpolation) other improvements as well.
- v1.1.6 - errors and warnings improved. smart adjustments for diffrent positioning style(of anchors elemntes and common
  ancestor element) . minor bug fixes

## 1.0.0

initial release.


src/utils/buzierSmooth.test.js
import { pointsToLines } from './buzierSmooth';

const points1 = [
  [5, 10],
  [10, 40],
  [40, 30],
  [60, 5],
  [90, 45],
  [120, 10],
  [150, 45],
  [200, 10],
];

const testPoints = () => {
  const points = [
    [0, 0],
    [10, 50],
  ];
  console.log(pointsToLines(points));
};

testPoints();

// console.log(svgPath(points,bezierCommand))


src/utils/index.ts
import { refType } from '../types';

export const getElementByPropGiven = (ref: refType): HTMLElement => {
  let myRef;
  if (typeof ref === 'string') {
    // myRef = document.getElementById(ref);
    myRef = document.getElementById(ref);
  } else myRef = ref.current;
  return myRef;
};

// receives string representing a d path and factoring only the numbers
export const factorDpathStr = (d: string, factor) => {
  let l = d.split(/(\d+(?:\.\d+)?)/);
  l = l.map((s) => {
    if (Number(s)) return (Number(s) * factor).toString();
    else return s;
  });
  return l.join('');
};

// debug
export const measureFunc = (callbackFunc: Function, name = '') => {
  const t = performance.now();

  const returnVal = callbackFunc();
  console.log('time ', name, ':', performance.now() - t);
  return returnVal;
};

// return relative,abs
export const xStr2absRelative = (str): { abs: number; relative: number } => {
  if (typeof str !== 'string') return { abs: 0, relative: 0.5 };
  console.log(str);
  let sp = str.split('%');
  let absLen = 0,
    percentLen = 0;
  if (sp.length == 1) {
    let p = parseFloat(sp[0]);
    if (!isNaN(p)) {
      absLen = p;
      return { abs: absLen, relative: 0 };
    }
  } else if (sp.length == 2) {
    let [p1, p2] = [parseFloat(sp[0]), parseFloat(sp[1])];
    if (!isNaN(p1)) percentLen = p1 / 100;
    if (!isNaN(p2)) absLen = p2;
    if (!isNaN(p1) || !isNaN(p2)) return { abs: absLen, relative: percentLen };
  }
};

export const getElemPos = (elem: HTMLElement) => {
  const pos = elem?.getBoundingClientRect() ?? { left: 0, top: 0, right: 0, bottom: 0 };
  return {
    x: pos.left,
    y: pos.top,
    right: pos.right,
    bottom: pos.bottom,
  };
};

// export const measureFunc = (func: Function, name = '') => (...args) => {
//   const t = performance.now();
//
//   console.log(this, func.name, ...args);
//   const returnVal = func(...args);
//   console.log('time ', func.name || name, ':', performance.now() - t);
//   return returnVal;
// };
//// example
// instead: myFunc(arg1,arg2)
// call: measureFunc(myFunc)(arg1,arg2)


src/utils/paths.js
// type pointType = {x:number,y:number}
//
// class Dir {
//   constructor(dir:pointType, label:string,reverse:boolean,clockwise:string) {
//
//   }
// }
// type pointType = {x:number,y:number}

const shortAnchor = (an) => {
  const d = {
    top: 'U',
    bottom: 'D',
    right: 'R',
    left: 'L',
  };
  return d[an];
};

export class Dir {
  static refs = {};

  constructor(dir, label, reverse, clockwise) {
    this.dir = dir;
    this.label = label;
    Dir.refs[label] = { ref: this, reverse, clockwise };
  }

  add = (root, add) => {
    return [root[0] + this.dir[0] * add, root[1] + this.dir[1] * add];
  };

  mul = (dir) => {
    return [this.dir[0] * dir[0], this.clockwise().dir[1] * dir[1]];
  };

  clockwise = () => Dir.refs[Dir.refs[this.label].clockwise].ref;

  reverse = () => Dir.refs[Dir.refs[this.label].reverse].ref;

  static get = (label) => Dir.refs[label].ref;
  static dirs = {
    U: new Dir([0, -1], 'U', 'D', 'R'),
    R: new Dir([1, 0], 'R', 'L', 'D'),
    D: new Dir([0, 1], 'D', 'U', 'L'),
    L: new Dir([-1, 0], 'L', 'R', 'U'),
  };
}

// console.log(Dir.dirs.top.reverse());

// start anchor direction
export const SAD = {
  top: 'U',
  right: 'R',
  bottom: 'D',
  left: 'L',
};
export const EAD = {
  top: 'D',
  right: 'L',
  bottom: 'U',
  left: 'R',
};

// console.log(Dir.dirs[SAD['top']]);

// const dirU = new Dir([0, -1], 'U', 'D', 'R');
// const dirR = new Dir([1, 0], 'R', 'L', 'D');
// const dirD = new Dir([0, 1], 'D', 'U', 'L');
// const dirL = new Dir([-1, 0], 'L', 'R', 'U');

// console.log(Dir.get('L').add([0, 0], 10));


src/utils/props.ts
import { xarrowPropsType } from '../types';

export const parseProps = (props: xarrowPropsType) => {
  let {
    startAnchor = 'auto',
    endAnchor = 'auto',
    label = null,
    color = 'CornflowerBlue',
    lineColor = null,
    headColor = null,
    tailColor = null,
    strokeWidth = 4,
    showHead = true,
    headSize = 6,
    showTail = false,
    tailSize = 6,
    path = 'smooth',
    curveness = 0.8,
    gridBreak = '50%',
    // gridRadius = strokeWidth * 2, //todo
    dashness = false,
    headShape = 'arrow1',
    tailShape = 'arrow1',
    showXarrow = true,
    animateDrawing = false,
    passProps = {},
    arrowBodyProps = {},
    arrowHeadProps = {},
    arrowTailProps = {},
    SVGcanvasProps = {},
    divContainerProps = {},
    divContainerStyle = {},
    SVGcanvasStyle = {},
    _extendSVGcanvas = 0,
    _debug = false,
    _pathMargin = 20,
    ...extraProps
  } = props;
};


src/types.ts
///////////////
// public types

// constants used for typescript and proptypes definitions
export const tAnchorEdge = ['auto', 'left', 'right', 'top', 'bottom', 'middle'] as const;
export const tPaths = ['smooth', 'grid', 'straight'] as const;
export const tSvgElems = ['circle', 'ellipse', 'line', 'path', 'polygon', 'polyline', 'rect'] as const;
export const tFacingDir = ['auto', 'inwards', 'outwards', 'left', 'right', 'up', 'down'] as const;

//default arrows svgs
export const arrowShapes = {
  arrow1: { svgElem: 'path', svgProps: { d: `M 0 0 L 1 0.5 L 0 1 L 0.25 0.5 z` }, offsetForward: 0.75 },
  heart: {
    svgElem: 'path',
    svgProps: {
      d: `M 0,0.25 A 0.125,0.125 0,0,1 0.5,0.25 A 0.125,0.125 0,0,1 1,0.25 Q 1,0.625 0.5,1 Q 0,0.625 0,0.25 z`,
    },
    offsetForward: 0.92,
  },
  circle: {
    svgElem: 'circle',
    svgProps: {
      r: 0.5,
      cx: 0.5,
      cy: 0.5,
      // children: <animate attributeName="r" values={'0.25;0.5;0.25'} dur="1s" repeatCount={'indefinite'} />,
    },
    offsetForward: 1,
  },
  test: { svgElem: 'path', svgProps: { d: `M 1 1 L 3 2 L 1 3 L 1.5 2 z` }, offsetForward: 1 },
  arrow2: {
    svgElem: 'path',
    svgProps: {
      //// handle automatic resize of the svg
      // d: `M 0.5 1 l -0.171749 -0.16666 0.3333 -0.3333 -0.3333 -0.3333 0.171749 -0.16666 0.494916 0.5 z`,
      // d: `M 0 1 l -0.171749 -0.16666 0.3333 -0.3333 -0.3333 -0.3333 0.171749 -0.16666 0.494916 0.5 z`,
      d: `M 0 24 l -4.122     -4      8      -8      -8      -8       4.122    -4 11.878 12 z`,
    },
    offsetForward: 0.4,
  },
  // todo: add more default shapes
} as const;

export const tArrowShapes = Object.keys(arrowShapes) as Array<keyof typeof arrowShapes>;

export type xarrowPropsType = {
  start: refType;
  end: refType;
  startAnchor?: anchorType;
  endAnchor?: anchorType;
  label?: labelType | labelsType;
  color?: string;
  lineColor?: string | null;
  headColor?: string | null;
  tailColor?: string | null;
  strokeWidth?: number;
  showHead?: boolean;
  headSize?: number;
  showTail?: boolean;
  tailSize?: number;
  path?: pathType;
  showXarrow?: boolean;
  curveness?: number;
  gridBreak?: string;
  // gridRadius?: number;
  dashness?:
    | boolean
    | {
        strokeLen?: number;
        nonStrokeLen?: number;
        animation?: boolean | number;
      };
  headShape?: svgEdgeShapeType | svgCustomEdgeType;
  tailShape?: svgEdgeShapeType | svgCustomEdgeType;
  animateDrawing?: boolean | number;
  avoidRects?: (refType | rectType)[];
  passProps?: JSX.IntrinsicElements[svgElemType];
  SVGcanvasProps?: React.SVGAttributes<SVGSVGElement>;
  arrowBodyProps?: React.SVGProps<SVGPathElement>;
  arrowHeadProps?: JSX.IntrinsicElements[svgElemType];
  arrowTailProps?: JSX.IntrinsicElements[svgElemType];
  divContainerProps?: React.HTMLProps<HTMLDivElement>;
  SVGcanvasStyle?: React.CSSProperties;
  divContainerStyle?: React.CSSProperties;
  _extendSVGcanvas?: number;
  _debug?: boolean;
  _pathMargin?: number;
};

export type pathType = typeof tPaths[number];
export type _anchorType = anchorNamedType | anchorCustomPositionType;
export type anchorType = _anchorType | _anchorType[];
export type anchorNamedType = typeof tAnchorEdge[number];

export type _faceDirType = typeof tFacingDir[number];
export type anchorCustomPositionType = {
  position: anchorNamedType;
  offset?: { rightness?: number; bottomness?: number };
  facingDir?: _faceDirType | _faceDirType[];
};
export type refType = React.MutableRefObject<any> | string;
export type labelsType = {
  start?: labelType;
  middle?: labelType;
  end?: labelType;
  [key: string]: labelType;
};
export type labelType = JSX.Element | string;

export type svgCustomTypeGeneric<T extends svgElemType> = {
  svgElem: T;
  svgProps?: JSX.IntrinsicElements[T];
  offsetForward?: number;
};
export type svgCustomEdgeType = { [K in svgElemType]: svgCustomTypeGeneric<K> }[svgElemType];
export type svgEdgeShapeType = typeof tArrowShapes[number];
export type svgEdgeType = svgEdgeShapeType | svgCustomEdgeType;
export type svgElemType = typeof tSvgElems[number];

////////////////
// private types

export type dimensionType = {
  x: number;
  y: number;
  right: number;
  bottom: number;
};
export type rectType = {
  x: number;
  y: number;
  width: number;
  height: number;
};
export type _prevPosType = {
  start: dimensionType;
  end: dimensionType;
};

export type anchorEdgeType = 'left' | 'right' | 'top' | 'bottom';
export type _xarrowVarPropsType = Omit<xarrowPropsType, 'start' | 'end'>;

// pick the common props between 2 objects
type Common<A, B> = {
  [P in keyof A & keyof B]: A[P] | B[P];
};

// const c: Common<T1, T2> = { y: 123 };

// export type SvgCustomTypeOneLiner = svgElemType extends infer T
//   ? T extends svgElemType
//     ? { svgElem: T; svgProps?: JSX.IntrinsicElements[T] }
//     : never
//   : never;

// export type svgCustomType<T extends svgElemType> = {
//   svgElem: T;
//   svgProps?: { [key in T]: JSX.IntrinsicElements[T] };
//   offsetForward?: number;
// };
